File: 프로그래머스/3/49191. 순위/순위.py
from collections import defaultdict

def solution(n, results):
    answer = 0
    
    win_graph = defaultdict(list) # 자신이 이긴 선수 리스트
    lose_graph = defaultdict(list) # 자신에게 이긴 선수 리스트
    
    for winner, loser in results:
        win_graph[winner].append(loser)
        lose_graph[loser].append(winner)
    
    def dfs(graph, start, visited): # start는 승패를 파악하고자 하는 선수
        for player in graph[start]:
            if player not in visited:
                visited.add(player)
                dfs(graph, player, visited)
    
    for player in range(1, n+1):
        win_set = set()
        lose_set = set()
        
        dfs(win_graph, player, win_set) # player가 이긴 선수들
        dfs(lose_graph, player, lose_set) # player가 진 선수들
        
        if len(win_set) + len(lose_set) == n - 1:
            answer += 1
    
    return answer

File: 프로그래머스/3/49191. 순위/순위.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] results) {
        int answer = 0;
        
        // key가 이긴 선수 리스트
        List<List<Integer>> winGraph = new ArrayList<>();
        // key를 이긴 선수 리스트
        List<List<Integer>> loseGraph = new ArrayList<>();
        
        // graphs 초기화
        for (int i = 0; i < n+1; i++) {
            winGraph.add(new ArrayList<>());
            loseGraph.add(new ArrayList<>());
        }
        
        for (int[] result: results) {
            int win = result[0], lose = result[1];
            winGraph.get(win).add(lose);
            loseGraph.get(lose).add(win);
        }
        
        for (int p = 1; p < n+1; p++) {
            Set<Integer> winVisited = new HashSet<>();
            Set<Integer> loseVisited = new HashSet<>();
            
            int winCount = dfs(p, winGraph, winVisited);
            int loseCount = dfs(p, loseGraph, loseVisited);
            
            // 승패를 모두 가려냄
            if (winCount + loseCount == n-1) {
                answer++;
            }
        }
        
        return answer;
    }
    
    static int dfs(int start, List<List<Integer>> graph, Set<Integer> visited) {
        int count = 0; // 이긴/진 선수 수
        for (Integer player: graph.get(start)) {
            if (!visited.contains(player)) {
                visited.add(player);
                count += 1 + dfs(player, graph, visited);
            }
        }
        return count;
    }
}